crates/sui-core/src/authority_active/execution_driver/mod.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use sui_types::{base_types::TransactionDigest, error::SuiResult, messages::CertifiedTransaction};
+use tracing::debug;
+use typed_store::Map;
+
+use crate::{authority::AuthorityState, authority_client::AuthorityAPI};
+
+use super::{gossip::LocalConfirmationTransactionHandler, ActiveAuthority};
+
+#[cfg(test)]
+pub(crate) mod tests;
+
+pub trait PendCertificateForExecution {
+    fn pending_execution(
+        &self,
+        certs: Vec<(TransactionDigest, CertifiedTransaction)>,
+    ) -> SuiResult<()>;
+}
+
+impl PendCertificateForExecution for AuthorityState {
+    fn pending_execution(
+        &self,
+        certs: Vec<(TransactionDigest, CertifiedTransaction)>,
+    ) -> SuiResult<()> {
+        self.database.add_pending_certificates(certs)
+    }
+}
+
+pub struct PendCertificateForExecutionNoop;
+impl PendCertificateForExecution for PendCertificateForExecutionNoop {
+    fn pending_execution(
+        &self,
+        _certs: Vec<(TransactionDigest, CertifiedTransaction)>,
+    ) -> SuiResult<()> {
+        Ok(())
+    }
+}
+
+/// When a notification that a new pending transaction is received we activate
+/// processing the transaction in a loop.
+pub async fn execution_process<A>(active_authority: &ActiveAuthority<A>)
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    debug!("Start pending certificates execution.");
+
+    // Loop whenever there is a signal that a new transactions is ready to process.
+    loop {
+        // NOTE: nothing terrible happens if we fire more often than there are
+        //       transactions awaiting execution, or less often than once per transactions.
+        //       However, we need to be sure that if there is an awaiting trasnactions we
+        //       will eventually fire the notification and wake up here.
+        active_authority.state.database.wait_for_new_pending().await;
+
+        debug!("Pending certificate execution activated.");
+
+        if let Err(err) = execute_pending(active_authority).await {
```

Comment:
What can trigger this error exactly?

Reply:
I think only own system error can trigger this -- disconnected from network, or storage error.

Reply:
I think it could possibly encounter timeouts under non-exotic conditions. Since sync_authority_source_to_destination syncs certs and their parents serially, the minimum time to finish syncing a cert is RTT to source authority * number of new certs in dependency graph. It doesn't take too much for that number to exceed 10 seconds. Whether this happens in practice or not depends on how far from causal order we stray.